### ISSUE TITLE
After deploying: `Error: [@octokit/webhooks] signature does not match event payload and secret`

### ISSUE THREAD

**Issue by mrousavy** [probot/example-vercel#82]
I developed my bot according to this example, but after deploying it to Vercel I get these errors in my Vercel Runtime Logs:

```
AggregateError: 
Error: [@octokit/webhooks] signature does not match event payload and secret
at verifyAndReceive (/var/task/node_modules/@octokit/webhooks/dist-node/index.js:490:19)
at processTicksAndRejections (node:internal/process/task_queues:95:5)
at middleware (/var/task/node_modules/@octokit/webhooks/dist-node/index.js:621:5)
at Server.<anonymous> (/opt/rust/nodejs.js:1:10461)
at Server.<anonymous> (/opt/rust/nodejs.js:7:3888)
at receiverHandle (/var/task/node_modules/@octokit/webhooks/dist-node/index.js:395:33)
at verifyAndReceive (/var/task/node_modules/@octokit/webhooks/dist-node/index.js:493:31)
at processTicksAndRejections (node:internal/process/task_queues:95:5)
at middleware (/var/task/node_modules/@octokit/webhooks/dist-node/index.js:621:5)
at Server.<anonymous> (/opt/rust/nodejs.js:1:10461)
at Server.<anonymous> (/opt/rust/nodejs.js:7:3888) {
event: Error: [@octokit/webhooks] signature does not match event payload and secret
at verifyAndReceive (/var/task/node_modules/@octokit/webhooks/dist-node/index.js:490:19)
at processTicksAndRejections (node:internal/process/task_queues:95:5)
at middleware (/var/task/node_modules/@octokit/webhooks/dist-node/index.js:621:5)
at Server.<anonymous> (/opt/rust/nodejs.js:1:10461)
at Server.<anonymous> (/opt/rust/nodejs.js:7:3888) {
```

Any idea why?

**Comment by skywarth** [comment]
Just my two cents, it might help:
- If this error is not present in your local, verify whether you install the dependencies via `npm install` or `npm ci`. If there is no error on local, this might mean you're installing with `npm install` which could essentially differ dependency versions and cause this error
- Do you have `PRIVATE_KEY`, `WEBHOOK_SECRET` and other similar `.env` values set in production? Verify them please.

**Comment by mrousavy** [comment]
Actually the problem was that Vercel was doing some verifications/parsing on webhook requests - I disabled that and now it works! Thanks

**Comment by jonasfroeller** [comment]
> Actually the problem was that Vercel was doing some verifications/parsing on webhook requests - I disabled that and now it works! Thanks

Where did you disable that? I am facing the same issue right now.

**Comment by skywarth** [comment]
Hi @jonasfroeller,

I've also faced the issue described by the OP. I fixed it by adding the following env variable in my Vercel configuration (on settings page):
`NODEJS_HELPERS=0`

Screenshot:
![image](https://github.com/user-attachments/assets/92f5fc17-2f15-466e-b54a-4143199e0b2e)

Let me know if it resolves it for you, I'm happy to help.

**Comment by jonasfroeller** [comment]
> Hi @jonasfroeller,
> 
> I've also faced the issue described by the OP. I fixed it by adding the following env variable in my Vercel configuration (on settings page): `NODEJS_HELPERS=0`
> 
> Screenshot: ![image](https://private-user-images.githubusercontent.com/23261263/356228436-92f5fc17-2f15-466e-b54a-4143199e0b2e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMxMjQxODgsIm5iZiI6MTcyMzEyMzg4OCwicGF0aCI6Ii8yMzI2MTI2My8zNTYyMjg0MzYtOTJmNWZjMTctMmYxNS00NjZlLWI1NGEtNDE0MzE5OWUwYjJlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA4VDEzMzEyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljMTgyMjMwOGQ4YzM0NjNhYzg5MmE0Y2YzNGEyNDU2N2FkYzI0ODFiMGY2MGQwZjRiNjk5MzcyZDJhOTVjMzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.POD9txGHjuepeFrA0O_PUp-ihEPrWi4OLso_8Xld39w)
> 
> Let me know if it resolves it for you, I'm happy to help.

Thank you for the quick response. It did not work unfortunately. Did I miss any environment variable needed for vercel?
![image](https://github.com/user-attachments/assets/7e578ca5-54d0-4e9a-8a76-1236a9b81657)

**Comment by skywarth** [comment]
@jonasfroeller sorry to hear that. Let's check it against my current env varibles in Vercel to see if anything is missing. See below:

![image](https://github.com/user-attachments/assets/2e7e7f12-9d06-417e-b24d-723bac5bd3f0)

It seems like all the required keys are present in your env. Can you check for these please:
- Does your application have any `.env` file in the repository, committed and present VCS?
- Does the project has any configuration or setup step which automatically creates a `.env` file?
- Could you verify the values of these env keys once again with your local env, beware of whitespaces and case sensitivity
- Is the currently nominal and working in your local environment via smee.io? 
- Can you share the most recent webhook payload details from GitHub app (Hide the secrets for security & privacy concerns)
  - Navigate to it via GitHub->Settings->Developer Settings->Edit (for the app)->Advanced
  - Response received, response code, headers etc. Payload is not really necessary
- Can you also share the received request on Vercel traffic monitoring (on Logs tab)

I suppose it's either 404, 400 or 500 error, blocking the webhook request sent by the GitHub. I also struggled while setting up the Vercel, took me quite some hours. As far as I remember I did that `NODEJS_HELPERS` trick and maybe some other stuff which I cannot recall right now. 

Also check my [app's vercel function definition](https://github.com/skywarth/darkest-PR/blob/master/api/github/webhooks/index.ts) and compare it to yours, it could also be impacting the deployment.

Let me know how it pans out.

**Comment by jonasfroeller** [comment]
@skywarth Thanks a lot.

* No, no `.env` in the project. Just a `.env.example`.
* No, the [repository](https://github.com/neptun-software/neptun.github.app) is currently basically just this probot example using ts and esm. No github workflows or postinstall scripts or anything like that (other than `npm run build` for ts).
* I checked this 3 times now. I also tried it with freshly generated secrets. I also tried creating a new github app from scratch. (and I removed all linebreaks from the private key)
* It works locally with smee.io, yes.
* Response i get (redelivery):

```
# Headers
Cache-Control: public, max-age=0, must-revalidate
Content-Length: 88
Content-Type: text/plain; charset=utf-8
Date: Thu, 08 Aug 2024 14:17:18 GMT
Server: Vercel
Strict-Transport-Security: max-age=63072000; includeSubDomains; preload
X-Vercel-Cache: MISS
X-Vercel-Id: iad1::dub1::dpxpj-xxx-xxx

# Body
{"error":"Error: [@octokit/webhooks] signature does not match event payload and secret"}

# Code
400
```

* Vercel logs:

![image](https://github.com/user-attachments/assets/ca357c51-5dfb-4c61-a3d4-258811a004fb)
```
  errors: [
    Error: [@octokit/webhooks] signature does not match event payload and secret
        at verifyAndReceive (/var/task/node_modules/@octokit/webhooks/dist-node/index.js:490:19)
        at middleware (/var/task/node_modules/@octokit/webhooks/dist-node/index.js:621:5)
        at Server.<anonymous> (/opt/rust/nodejs.js:8:3933) {
      event: [Object],
      status: 400
    }
  ]
```

**Comment by skywarth** [comment]
Ah, glad it works now (if I understood correctly). Yes I was about to comment whether you tried to redeploy (e.g: trigger via pushing to repository) **after** you set the `NODEJS_HELPERS=0` env variable. Because new env usually takes effect after the app is redeployed on Vercel. 

Oh and don't forget to remove references to smee.io from Vercel env if there is any, smee.io is for local debugging and development only.  

After inspecting the input you've given, I can confirm it was the exact same problem I had trouble with in the past. 

Leaving here some additional resources in case someone stumbles here in the future:
- https://probot.github.io/docs/deployment/#vercel
- https://vercel.com/docs/functions/runtimes/node-js/advanced-node-configuration#disabling-helpers-for-node.js

Ping me up if you need anything else, I'm glad to assist. Cheers.

**Comment by jonasfroeller** [comment]
@skywarth I have 2 Github apps. One for development and one for production, so that I do not have to change the webhook url all the time (yes, both of them work now). The weird thing is that I redeployed everytime, that I changed the env variables, but it only worked now, when I did not actually update anything. Really strange. Maybe there was a Github or Vercel incident.

Thanks again! I think I would have just gave up and used docker, if nobody responded here.
